### PR TITLE
Optimize addToTree function.

### PR DESCRIPTION
**Pull request by yurydelendik**
Associated Issue: #4168

### Summary of Changes

* don't use sortTree  for sources
* improve the performance of addToTree

### Test Plan

Added console.time() in the SourcesTree.js and createTree.js around addToTree+sortTree operations.
Measured performance of https://yurydelendik.github.io/old-man-sandbox/create-30k/f3k.html tree.

Before:
screen shot 2017-10-03 at 2 59 13 pm

After:
screen shot 2017-10-03 at 3 03 50 pm

Performance improved from 1186ms->291ms also affected subsequent/incremental tree update operations 180ms->2ms.

### CODE REVIEW

File: src/utils/sources-tree/addToTree.js [hunk 1]
```diff
@@ -1,5 +1,7 @@
 // @flow
 
+import { parse } from "url";
```
Comment (jasonLaster) [inline src/utils/sources-tree/addToTree.js]: does it make sense to have this in `getURL` to have one module for URL parsing?

File: src/utils/sources-tree/addToTree.js [hunk 2]
```diff
@@ -12,14 +14,118 @@ import { getURL, getFilenameFromPath } from "./getURL";
...
+ */
+function isExactDomainMatch(part: string, debuggeeHost: string): boolean {
+  return part.indexOf("www.") === 0
+    ? part.substr(4) === debuggeeHost
+    : part === debuggeeHost;
+}
+
+type FindNodeInContentsMatcher = any => number;
+/*
+ * Performs a binary search to insert a node into contents. Returns positive
```
Comment (jasonLaster) [inline src/utils/sources-tree/addToTree.js]: :+1:

File: src/utils/sources-tree/addToTree.js [hunk 3]
```diff
@@ -12,14 +14,118 @@ import { getURL, getFilenameFromPath } from "./getURL";
...
+    return null;
+  }
+  return host.indexOf("www.") === 0 ? host.substr(4) : host;
+}
+
+/*
+ * Checks if node name matches debugger host/domain.
+ */
+function isExactDomainMatch(part: string, debuggeeHost: string): boolean {
+  return part.indexOf("www.") === 0
```
Comment (adrian17) [inline src/utils/sources-tree/addToTree.js]: just curious: why not `part.startsWith("www.")` ?

File: src/utils/sources-tree/addToTree.js [hunk 4]
```diff
@@ -12,14 +14,118 @@ import { getURL, getFilenameFromPath } from "./getURL";
...
+
+type FindNodeInContentsMatcher = any => number;
+/*
+ * Performs a binary search to insert a node into contents. Returns positive
+ * number, index of the found child, or negative number, which can be used
+ * to calculate a position where a new node can be inserted (`-index - 1`).
+ * The matcher is a function that returns result of comparision of a node with
+ * lookup value.
+ */
+function findNodeInContents(tree: Node, matcher: FindNodeInContentsMatcher) {
```
Comment (codehag) [inline src/utils/sources-tree/addToTree.js]: describing this function in plain english, and using the ternary values I defined elsewhere (`T_TRUE = 1, T_UNKNOWN = 0, T_FALSE = -1`):

* we take the tree contents which is an array. if that array is empty, we return -1.

* if it is not empty, we set a start point ('l' for left) and an end point ('r' for right)

* so long as the starting point is less than the end point, we also select a center point ('c')

* if the center point is `T_FALSE` ( -1 ) -- then it should be at the bottom of the list, so we set the start point (l) to c + 1

* otherwise, the center point is `T_UNKNOWN`, or `T_TRUE`, so we set the ending point to C

example:

```
0|1|2|3|4
----------
l| |c| |r

if (matcher(contents[c]) === `T_FALSE`) 
0|1|2|3|4
----------
 | | |l, c|r // ignore the first half of the array

else
0|1|2|3|4
----------
l|c|r| |  // ignore the second half of the array
```

* Once the loop completes, (and our starting point "l" is greater than our endpoint "r") 

```
0|1|2|3|4
----------
 |r|l| |
0|1|2|3|4
----------
 | | | |r|l // outside of the original array if all matcher checks return `T_FALSE`
```

we check if the result is `T_UNKNOWN`. If it is unknown, we return the starting value ("l").

If it is either `T_FALSE` (-1) or `T_TRUE` (1), then 

- check if it is `T_FALSE`, if it is, return the starting point ('l'). 
- if it is `T_TRUE`, return the starting point ('l') plus 1. 

whatever the result is, make it negative and -1 from it.

```
0|1|2|3|4
----------
 |r|l|x|   // if `T_FALSE` is true, the returned position will be -(l + 1) ( also see above for the off by one case, where  | | | |r|l|x -- is outside of the existing array by two spots)
0|1|2|3|4
----------
 |r|l| |x| // if `T_TRUE` is true, the returned position will be -(l + 2)
```

this code is essentially taking an array, seeing if the matching position would be above or below the center, and halving it to the side where the element might be found. That isnt immediately clear from the code. there are two things contributing to this: 

1) bad naming. Single character variable names shouldn't be used except in special situations, and this isn't one of them.
2) return value is doing too much. it is unclear to me why we need it to be negative, or why we need to add +1 to it. can you clarify why this is happening? also, there are a lot of magic things happening, for example +1 is being added in a number of places without any reason explained behind it. return value is a good example of "why are we adding 1 and then making the whole thing negative and then adding 1 to it again". at the very least the return value needs to be split up


it also seems like we are avoiding complex data structures, but in this case i think we need them. we shouldnt be checking for negative indexs to tell if a node is new or not.

File: src/utils/sources-tree/addToTree.js [hunk 5]
```diff
@@ -12,14 +14,118 @@ import { getURL, getFilenameFromPath } from "./getURL";
...
+
+/*
+ * Creates a matcher for findNodeInContents.
+ * The sorting order of nodes during comparison is:
+ * - "(index)" node
+ * - root node with the debuggee host/domain
+ * - hosts/directories (not files) sorted by name
+ * - files sorted by name
+ */
+function createTreeNodeMatcher(
```
Comment (codehag) [inline src/utils/sources-tree/addToTree.js]: this should be moved to its own file, each HOF can be split into its own function and named. that way you wont have to add so many comments

```javascript
function indexMatcher(indexName) {
  return (node) => (node.name === IndexName ? 0 : 1);
}
function domainMatcher(indexName, debuggeeHost) {
 return node => {
   if (node.name === IndexName) {
      return -1;
    }
    return isExactDomainMatch(node.name, (debuggeeHost: any)) ? 0 : 1;
  };
}

createTreeNodeMatcher(...) {
  //...
  if (part === IndexName) {
    return IndexMatcher(indexName)
  } 
  if (debuggeeHost && isExactDomainMatch(part, debuggeeHost)) {
    return domainMatcher(indexName, debuggeeHost)
  }
  // ... etc
```

File: src/utils/sources-tree/addToTree.js [hunk 6]
```diff
@@ -12,14 +14,118 @@ import { getURL, getFilenameFromPath } from "./getURL";
...
+  return part.indexOf("www.") === 0
+    ? part.substr(4) === debuggeeHost
+    : part === debuggeeHost;
+}
+
+type FindNodeInContentsMatcher = any => number;
+/*
+ * Performs a binary search to insert a node into contents. Returns positive
+ * number, index of the found child, or negative number, which can be used
+ * to calculate a position where a new node can be inserted (`-index - 1`).
```
Comment (codehag) [inline src/utils/sources-tree/addToTree.js]: so, if an index is negative, it is a new node? and if it is positive it is an existing node? that is difficult to understand. why not use a tuple? { index: <number>, newNode: <boolean> }

File: src/utils/sources-tree/addToTree.js [hunk 7]
```diff
@@ -31,23 +137,28 @@ function findOrCreateNode(
...
+  const childIndex = findNodeInContents(
+    subTree,
+    createTreeNodeMatcher(part, !addedPartIsFile, debuggeeHost)
+  );
 
   // we create and enter the new node
-  if (!child) {
-    return createNodeInTree(part, path, subTree);
+  if (childIndex < 0) {
+    return createNodeInTree(part, path, subTree, -childIndex - 1);
```
Comment (codehag) [inline src/utils/sources-tree/addToTree.js]: see other comments

File: src/utils/sources-tree/addToTree.js [hunk 8]
```diff
@@ -84,18 +204,24 @@ function addSourceToNode(node: Node, url: Object, source: SourceRecord) {
...
+    const existingNode = node.contents[childIndex];
     existingNode.contents = source;
     return node.contents;
   }
 
   // if this is a new file, add the new file;
   const newNode = createNode(name, source.get("url"), source);
-  return [...node.contents, newNode];
+  const contents = node.contents.slice(0);
+  contents.splice(-childIndex - 1, 0, newNode);
```
Comment (codehag) [inline src/utils/sources-tree/addToTree.js]: it is so unclear what is going on here. we shouldn't be using negative numbers to determine if something is a new node

File: src/utils/sources-tree/addToTree.js [hunk 9]
```diff
@@ -12,14 +14,118 @@ import { getURL, getFilenameFromPath } from "./getURL";
...
+  }
+  const result = matcher(contents[l]);
+  if (result === 0) {
+    return l;
+  }
+  return -(result > 0 ? l : l + 1) - 1;
+}
+
+/*
+ * Creates a matcher for findNodeInContents.
```
[the remark this answers is not in the document]
Comment (yurydelendik) [reply src/utils/sources-tree/addToTree.js]: Actually, it's similar to what any comparison function will return: negative for less, 0 for equial and positive for greater

File: src/utils/sources-tree/addToTree.js (hunk 4 above)
Comment (yurydelendik) [reply src/utils/sources-tree/addToTree.js]: Sorry, I don't follow your analysis. The algorithm above is basic binary search algorithm. I renamed 'l' and 'r' to 'left' and 'right' for clarity.

File: src/utils/sources-tree/addToTree.js [hunk 10]
```diff
@@ -12,14 +14,118 @@ import { getURL, getFilenameFromPath } from "./getURL";
...
+    if (debuggeeHost && isExactDomainMatch(node.name, debuggeeHost)) {
+      return -1;
+    }
+    const nodeIsDir = nodeHasChildren(node);
+    if (nodeIsDir && !isDir) {
+      return -1;
+    } else if (!nodeIsDir && isDir) {
+      return 1;
+    }
+    return node.name.localeCompare(part);
```
[the remark this answers is not in the document]
Comment (yurydelendik) [reply src/utils/sources-tree/addToTree.js]: Added comments to the FindNodeInContentsMatcher instead.

File: src/utils/sources-tree/addToTree.js (hunk 8 above)
Comment (yurydelendik) [reply src/utils/sources-tree/addToTree.js]: Changed return type to return an object (I hope JIT will not create lots of objects in memory to GC them later). Measured performance for 3k file -- there was no real change in the performance.

File: src/utils/sources-tree/addToTree.js (hunk 3 above)
Comment (yurydelendik) [reply src/utils/sources-tree/addToTree.js]: No reason, probably just a habit to support old browsers. Changed to startsWith for readability, since all browsers we support have it.

File: src/utils/sources-tree/addToTree.js (hunk 4 above)
Comment (yurydelendik) [reply src/utils/sources-tree/addToTree.js]: > describing this function in plain english, and using the ternary values I defined elsewhere

The algorithm is using return values similar to comparefn defined in http://www.ecma-international.org/ecma-262/5.1/#sec-15.4.4.11 , and these are not compatible with  ternary values.

File: src/utils/sources-tree/addToTree.js (hunk 1 above)
Comment (yurydelendik) [reply src/utils/sources-tree/addToTree.js]: Not at the moment -- I did not want to affect any other logic that might be in "assumed" in the code.

File: src/utils/sources-tree/addToTree.js (hunk 5 above)
Comment (yurydelendik) [reply src/utils/sources-tree/addToTree.js]: Code was moved to the ./treeOrder.js file.

File: src/utils/sources-tree/addToTree.js (hunk 4 above)
Comment (codehag) [reply src/utils/sources-tree/addToTree.js]: :+1: i mean ternary as in: https://en.wikipedia.org/wiki/Three-valued_logic -- which i see as being similar to what is going on in the sort / comparefn that you linked (the difference being that 0 is used for "equal" instead of "undefined"). But i could be wrong... anyway, thank you for the changes it reads much better!

File: src/utils/sources-tree/treeOrder.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+      left = middle + 1;
+    } else {
+      right = middle;
+    }
+  }
+  const result = matcher(contents[left]);
+  if (result === 0) {
+    return { found: true, index: left };
+  }
+  return { found: false, index: result > 0 ? left : left + 1 };
```
Comment (codehag) [inline src/utils/sources-tree/treeOrder.js]: thanks for clarifying why you were doing what you were doing before! The garbage collection might be an issue, true. I hadn't thought of that, and you are totally right

Perhaps a way to make the code readable and more memory efficient (trying to balance things :sweat_smile:), we could keep the code you had (with a comment about what it was doing, a more explicit comment)

then where the value is used, have a wrapper function that translates the value? 

ie)

```javascript
export function findNodeInContents(
  // ...
  // returns a negative value for a new node, offset by one
  // returns a positive value for an existing node
  return -(result > 0 ? left : left + 1 ) - 1;
```

with helpers
```javascript
function childFound(child) {
  return child > -1
} 

function determineIndex(child) {
  // return a valid index whether the child is new (an therefore a negative number off by one)
  // or if the child is an existing index
  return child > -1 ? child : -child - 1;
}
```


then where it is used:
```javascript
if (!childFound(childIndex)) {
  return createNodeInTree(part, path, subTree, determineIndex(childIndex));
}
```

File: src/utils/sources-tree/addToTree.js (hunk 4 above)
Comment (yurydelendik) [reply src/utils/sources-tree/addToTree.js]: there is no guarantee that localeCompare() will return only -1, 0, or 1 (based on spec). Writing `const result = name1.localeCompare(name2); return result === 0 ? T_UNKNOWN : result < 0 ? T_FALSE : T_TRUE;` to accommodate ternary values does not make any sense (and unreadable) IMHO.

**Comment by codehag** [comment]
Thanks for the clarifications yury! i made one more comment for your feedback

File: src/utils/sources-tree/treeOrder.js (hunk 11 above)
Comment (yurydelendik) [reply src/utils/sources-tree/treeOrder.js]: I would keep that as is for now for readability. We can re-open the issue later. It's hard to measure atm -- it's impossible to run it with 10,000 or 30,000 load.

**Comment by codehag** [comment]
awesome work!
